src/coord/src/coord.rs

```diff
@@ -2758,135 +2764,185 @@ impl Coordinator {
+                        .into_nontemporal()
+                        .map_err(|_e| {
+                            crate::error::CoordError::Unstructured(::anyhow::anyhow!(
+                                "OneShot plan has temporal constraints"
+                            ))
+                        })?;
+                    // We should only get excited if we can track down an index for `id`.
+                    // If `keys` is non-empty, that means we think one exists.
+                    for (index_id, (desc, _typ)) in dataflow_plan.index_imports.iter() {
+                        if let Some((key, val)) = key_val {
+                            if Id::Global(desc.on_id) == *id && &desc.keys == key {
+                                // Indicate an early exit with a specific index and key_val.
+                                fast_path = Some(Err((
+                                    *index_id,
+                                    Some(val.clone()),
+                                    map_filter_project.clone(),
+                                )))
+                            }
+                        } else if Id::Global(desc.on_id) == *id {
+                            // Indicate an early exit with a specific index and no key_val.
+                            fast_path = Some(Err((*index_id, None, map_filter_project.clone())))
+                        }
+                    }
+                }
+                // nothing can be done for non-trivial expressions.
+                _ => {}
+            }
+        }
+
+        // There are three cases going forward, based on the variants of `fast_path`:
+        // 1. `Some(Ok(rows))` indicates a constant expression that we can return.
+        // 2. `Some(Err(id, key, mfp))` indicase an index read from `id`, using an optional `key`, with a `mfp` to push.
+        // 3. `None` means that `dataflow_plan` must be installed, and then peeked.
+        // The last two cases have a lot of code in common, and are lumped together.
+        let resp = if let Some(Ok(rows)) = fast_path {
+            // If the dataflow optimizes to a constant expression, we can immediately return the result.
+            let mut rows = match rows {
                 Ok(rows) => rows,
-                Err(e) => return Err(e.clone().into()),
+                Err(e) => return Err(e.into()),
             };
+            // retain exactly those updates less or equal to `timestamp`.
+            for (_, time, diff) in rows.iter_mut() {
+                if time.less_equal(&timestamp) {
+                    // clobber the timestamp, so consolidation occurs.
+                    *time = timestamp.clone();
+                } else {
+                    // zero the difference, to prevent a contribution.
+                    *diff = 0;
+                }
+            }
+            // Consolidate down the results to get correct totals.
+            differential_dataflow::consolidation::consolidate_updates(&mut rows);
```

Comment:
It's unfortunate that this logic and the below row count cloning happen in the coordinator thread. Still ok with this, but I may follow up with the change we discussed previously about having dataflow communicate (row, diff) and then moving the row count expansion work to the .fold call below because i think that runs in the session's tokio task.

Reply:
I'm up for making that change, but I think it has some knock on consequences about how quickly dataflow can return if it sees negative counts (if it only ships diffs, should it error if it sees a negative, or allow the recipient to own that). But, still prefer that.

Reply:
I think having dataflow just ship everything to coord, which makes error decisions, is good.